File: dsa-problems/Greedy/solutions/java/GRD-005-shuttle-overtime-minimizer.java
import java.util.*;

class Solution {
    public long minOvertimeCost(int n, long H, int[][] shifts) {
        long totalStandard = 0;
        int minRate = Integer.MAX_VALUE;
        
        for (int[] shift : shifts) {
            totalStandard += shift[0];
            if (shift[1] < minRate) {
                minRate = shift[1];
            }
        }
        
        if (totalStandard >= H) {
            return 0;
        }
        
        long needed = H - totalStandard;
        return needed * minRate;
    }
}

public class Main {
    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        if (!sc.hasNextInt()) return;
        
        int n = sc.nextInt();
        long H = sc.nextLong();

        int[][] shifts = new int[n][2];
        for (int i = 0; i < n; i++) {
            shifts[i][0] = sc.nextInt();
            shifts[i][1] = sc.nextInt();
        }

        Solution solution = new Solution();
        System.out.println(solution.minOvertimeCost(n, H, shifts));
        sc.close();
    }
}

File: dsa-problems/Greedy/solutions/cpp/GRD-005-shuttle-overtime-minimizer.cpp
#include <iostream>
#include <vector>
#include <algorithm>

using namespace std;

class Solution {
public:
    long long minOvertimeCost(int n, long long H, vector<pair<int,int>>& shifts) {
        long long totalStandard = 0;
        int minRate = 2e9; // Initialize with a large value
        
        for (const auto& shift : shifts) {
            totalStandard += shift.first;
            if (shift.second < minRate) {
                minRate = shift.second;
            }
        }
        
        if (totalStandard >= H) {
            return 0;
        }
        
        long long needed = H - totalStandard;
        return needed * minRate;
    }
};

int main() {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);

    int n;
    long long H;
    if (!(cin >> n >> H)) return 0;

    vector<pair<int,int>> shifts(n);
    for (int i = 0; i < n; i++) {
        cin >> shifts[i].first >> shifts[i].second;
    }

    Solution solution;
    cout << solution.minOvertimeCost(n, H, shifts) << "\n";

    return 0;
}
